File: easy/201-300/237/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use regex::Regex;

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

fn find_longest_word(letters: &str, dictionary: Vec<String>) -> String {
    let mut longest_word = String::new();
    let pattern = Regex::new(format!("^[{}]+$", letters).as_str()).unwrap();
    for word in dictionary {
        if pattern.is_match(word.as_str()) && word.len() > longest_word.len() {
            longest_word = word.to_string();
        }
    }
    longest_word
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stub() {
        assert_eq!(
            "",
            find_longest_word(
                "xxxx",
                vec![
                    "be".to_string(),
                    "bee".to_string(),
                    "ebb".to_string(),
                    "cab".to_string(),
                    "ghost".to_string()
                ]
            )
        );
        assert_eq!(
            "bee",
            find_longest_word(
                "eb",
                vec![
                    "be".to_string(),
                    "bee".to_string(),
                    "ebb".to_string(),
                    "cab".to_string(),
                    "ghost".to_string()
                ]
            )
        );
    }
}

File: easy/201-300/237/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"regexp"
	"strings"
)

var zPrint = fmt.Println
var dictionaryPath = filepath.Join("..", "enable1.txt")

func main() {
	_, _ = zPrint("hello world")
}

func loadDictionary(filename string) []string {
	byteContents, _ := ioutil.ReadFile(filename)
	return strings.Split(
		strings.Trim(string(byteContents), "\n"),
		"\n",
	)
}

func findLongestWordGivenLetters(letters string, dictionary []string) (longestWord string) {
	wordPattern := regexp.MustCompile(fmt.Sprintf("^[%s]+$", letters))
	for _, word := range dictionary {
		if len(word) > len(longestWord) && wordPattern.MatchString(word) {
			longestWord = word
		}
	}
	return
}
